Bluetooth: Controller: Fix missing connection handle invalidate

cvinayak opened the pull request:
Fix missing connection handle invalidate on Controller power up.

The connection context are zero-initialized on startup and calls to `ll_connected_get()` would incorrectly return a valid connection context pointer for connection handle 0.

cvinayak committed:
Bluetooth: Controller: Fix missing connection handle invalidate

Fix missing connection handle invalidate on Controller
power up.

The connection context are zero-initialized on startup and
calls to `ll_connected_get()` would incorrectly return a
valid connection context pointer for connection handle 0.

Relates to commit fa02dc4d02fd ("Bluetooth: Controller: Fix
missing reset of connection handle").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1667,6 +1668,11 @@ static int init_reset(void)
 
+	/* Invalidate connection handles, refer to ll_connected_get() */
+	for (handle = 0U; handle < CONFIG_BT_MAX_CONN; handle++) {
+		conn->lll.handle = LLL_HANDLE_INVALID;
```
Fixed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1836,7 +1845,6 @@ static inline void disable(uint16_t handle)
 					conn, &conn->lll);
 	LL_ASSERT_INFO2(err == 0 || err == -EALREADY, handle, err);
 
-	conn->lll.handle = LLL_HANDLE_INVALID;
```
> 1. Would this not make sure the handles are invalid at startup via `ll_reset()`? Or is that not always called on startup? I assume we would have seen a lot of errors if this issue is always present...

The issue is related to use of `hci_uart` without use of HCI Reset Command, and control path leading to call to `ll_tx_mem_enqueue()`. This is not observed when using Zephyr Host that uses HCI reset at startup.






> 2\. Even if it doesn't work at startup, shouldn't it be kept for later calls to `ll_reset`, ie. later BT_HCI_OP_RESET commands?

The assignment is now in `init_reset()` which is called in both `ull_conn_init()` and `ull_conn_reset()`; the handle is invalidated for both startup and use of HCI reset.

cvinayak commented:
Updated commit message to refer to https://github.com/cvinayak/zephyr/commit/fa02dc4d02fd74c7754f0f620e27e7b54329d7dd

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1667,6 +1668,14 @@ static int init_reset(void)
+	for (uint16_t handle = 0U; handle < CONFIG_BT_MAX_CONN; handle++) {
+		struct ll_conn *conn;
+
+		conn = ll_conn_get(handle);
+		conn->lll.handle = LLL_HANDLE_INVALID;
```
This seems to assume that the handles are already initialized to some value since you are calling `ll_conn_get` with handle 0, 1, 2, etc.

cvinayak replied:
`ll_conn_get()` translate to being array indexing of `conn_pool`.

https://github.com/zephyrproject-rtos/zephyr/blob/a1ba2249597b253b4eae0873dbef4677473a6fc2/subsys/bluetooth/controller/util/mem.c#L102

Thalley replied:
Ah, I see. That's pretty misleading this case, as handle != index in many cases. 

If the `ll_conn_get` function at some point will be modified to not use handle as an index, then this will (silently?) break. 

There is one other place where handles are treat as indexes: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/ll_sw/ull_peripheral_iso.c#L67-L76

I think neither that, nor this, should treat handle as indexes, unless we can have some additional check in place to ensure this. 

Alternatively, we could define a `ll_conn_get_by_index` to take an index, rather than a handle.

cvinayak replied:
> If the ll_conn_get function at some point will be modified to not use handle as an index, then this will (silently?) break.

Currently, all `handle` generated/produced by the Controller are indexes. Any handle  provided by Host, example, advertising handle is converted to `handle` which are indexes internally.

There is no plan to not use indexes internally, to keep the handle to context de-referencing at O(1).

cvinayak replied:
Updated `ll_conn_get()` to check and return `NULL`. Added `LL_ASSERT` in callers; The commit has increased the full-feature code size by ~1.2 KB!.

These assertion checks are development checks on handle and connection context values that are verified at the first reference; I will try classify such assertion checks as conditionally compilable asserts under `CONFIG_ASSERT=y` so that a reduce code size Controller can be built for target testing (All assertion check be enabled for BabbleSIM CI testing).

cvinayak committed:
Bluetooth: Controller: Add ll_conn_get() return value check

Add ll_conn_get() return value check for valid connection
context.

Build command:
cmake -GNinja -DBOARD=nrf52833dk/nrf52833
 -DEXTRA_CONF_FILE=overlay-all-bt_ll_sw_split.conf
 -DDTC_OVERLAY_FILE=boards/nrf52833dk_nrf52833_df.overlay
 -DSNIPPET="bt-ll-sw-split" ../../samples/bluetooth/hci_uart
ninja

Before:
Memory region         Used Size  Region Size  %age Used
           FLASH:      283716 B       512 KB     54.11%
             RAM:      109752 B       128 KB     83.73%
        IDT_LIST:          0 GB        32 KB      0.00%

After:
Memory region         Used Size  Region Size  %age Used
           FLASH:      284992 B       512 KB     54.36%
             RAM:      109752 B       128 KB     83.73%
        IDT_LIST:          0 GB        32 KB      0.00%

After (use of `conn != NULL`):
Memory region         Used Size  Region Size  %age Used
           FLASH:      285044 B       512 KB     54.37%
             RAM:      109752 B       128 KB     83.73%
        IDT_LIST:          0 GB        32 KB      0.00%

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Remove conn context NULL check

Remove conn variable NULL check before calling
ull_cp_release_tx() to support conditional compilation of
LLCP_TX_CTRL_BUF_QUEUE_ENABLE variant.

Relates to commit 1ff458ec875c ("Bluetooth: controller:
llcp: fixing tx buffer queue handling").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -180,19 +180,19 @@ uint16_t ll_conn_handle_get(struct ll_conn *conn)
-
 	conn = ll_conn_get(handle);
-	if (conn->lll.handle != handle) {
+	if (!conn || conn->lll.handle != handle) {
```
```suggestion
	if (conn == NULL || conn->lll.handle != handle) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -804,6 +807,9 @@ struct lll_conn *ull_conn_lll_get(uint16_t handle)
 	struct ll_conn *conn;
 
 	conn = ll_conn_get(handle);
+	if (!conn) {
```
```suggestion
	if (conn == NULL) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_central_iso.c:
```diff
@@ -155,6 +155,7 @@ static int prepare_cb(struct lll_prepare_param *p)
 
 	/* Get reference to ACL context */
 	conn_lll = ull_conn_lll_get(cis_lll->acl_handle);
+	LL_ASSERT(conn_lll);
```
```suggestion
	LL_ASSERT(conn_lll != NULL);
```

Applies many places

Thalley replied:
It's been in our coding guidelines for quite a while actually: Rule 85 of https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html: https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_14_04.c

cvinayak replied:
I added a commit and a revert commit as there is code size increase in use of explicit comparison.

The assertion check is correct to have, but the explicit checks seems an overhead due to coding style (both Bluetooth subsystem and few kernel code break this style anyway).

I have come to realization that these invalid checks like LL_ASSERT; and explicit comparisons total ~15 KB of code size on Cortex-M CPUs in the Controller hci_uart builds!

Reference exercise: #90599

Thalley replied:
That's surprising :o Do you know why changing from e.g. `conn_lll)` to `conn_lll != NULL` increase the code size?

cvinayak replied:
Apparently, on investigation, the assert string includes the conditional as string! ... i will put back the MISRA coding style.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1667,6 +1683,15 @@ static int init_reset(void)
 		 sizeof(conn_pool) / sizeof(struct ll_conn), &conn_free);
 
+	/* Invalidate connection handles, refer to ll_connected_get() */
+	for (uint16_t handle = 0U; handle < CONFIG_BT_MAX_CONN; handle++) {
```
```suggestion
	for (int handle = 0U; handle < CONFIG_BT_MAX_CONN; handle++) {
```

Suggest to use same type for the iterator and the value compared to

Thalley replied:
Fair :) We don't have a rule for this IIRC

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1667,6 +1683,15 @@ static int init_reset(void)
+		struct ll_conn *conn;
+
+		/* handle in valid range, conn will be non-NULL */
+		conn = ll_conn_get(handle);
+		conn->lll.handle = LLL_HANDLE_INVALID;
```
The changes in this PR implies that `ll_conn_get` can now return NULL, which is missing a check here before accessing the pointer on the next line

Thalley reviewed (dismissed):
Will approve for now to get this fix in. The code size and coding guidelines issues can be further investigated outside this PR

cvinayak commented:
> Will approve for now to get this fix in. The code size and coding guidelines issues can be further investigated outside this PR

On some more investigation, the code size increase is due to the assert string including the conditional (not due to explicit comparison!) . I will go back to using MISRA coding guidelines in this PR. The code size increase is acceptable.
